File: solutions/haskell/proverb/1/src/Proverb.hs
module Proverb (recite) where

import Text.Printf (printf)

recite :: [String] -> String
recite [] = ""
recite pieces@(first : rest) = init . unlines $ zipWith (curry verse) pieces rest ++ [ending]
 where
  verse (x, y) = printf "For want of a %s the %s was lost." x y
  ending = printf "And all for the want of a %s." first

File: solutions/clojure/proverb/1/src/proverb.clj
(ns proverb
  (:require [clojure.string :as str]))

(defn recite [items]
  (if (seq items)
    (str/join "\n"
              (conj (mapv #(format "For want of a %s the %s was lost." %1 %2)
                          items (rest items))
                    (format "And all for the want of a %s." (first items))))
    ""))
